### PR TITLE
celinspector: fix variable scope conflicts and resource detection edge cases

### PR DESCRIPTION
**Pull request by a-hilaly**
Fix several edge cases in the CEL inspector around variable scoping and resource detection:

- Fix variable scope conflicts between top-level declarations and loop variables
  (e.g when 'i' is both a declared resource and used as a loop variable in filter)
- Add proper tracking of loop variables to avoid false positives in unknown resource detection
- Fix chain method calls on unknown resources to correctly track the dependency chain
- Handle ambiguous identifiers more robustly in nested scopes
- Add comprehensive test coverage for edge cases and variable conflicts

The changes make the inspector more reliable when dealing with complex expressions
that mix resource references with loop variables and method chains.
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: internal/typesystem/celinspector/celinspector.go [hunk 1]
```diff
@@ -80,9 +129,12 @@ func NewInspector(resources []string, functions []string) (*Inspector, error) {
 		env:       env,
 		resources: resourceMap,
 		functions: functionMap,
+		loopVars:  make(map[string]struct{}),
 	}, nil
 }
 
+// NewInspectorWithEnv creates a new Inspector instance with the given resources and functions
+// using the provided CEL environment.
```
Comment (michaelhtm) [inline internal/typesystem/celinspector/celinspector.go:137]: Should we add loopVars here too?

File: internal/typesystem/celinspector/celinspector.go [hunk 2]
```diff
@@ -202,88 +278,219 @@ func (a *Inspector) inspectIdent(ident *exprpb.Expr_Ident, currentPath string) E
...
+	case *exprpb.Constant_NullValue:
+		return "null"
+	default:
+		return "<unknown constant>"
+	}
+}
+
+// callToString converts a function call expression to its string representation.
+// This includes both regular function calls and operator calls.
+func (a *Inspector) callToString(call *exprpb.Expr_Call) string {
```
Comment (michaelhtm) [inline internal/typesystem/celinspector/celinspector.go]: nit: could this be called `callExpressionToString`?

File: internal/typesystem/celinspector/celinspector.go [hunk 3]
```diff
@@ -133,45 +193,63 @@ func (a *Inspector) inspectAst(expr *exprpb.Expr, currentPath string) Expression
...
+		// This is an unknown function, but not an internal one
+		inspection.UnknownFunctions = append(inspection.UnknownFunctions, UnknownFunction{Name: call.Function})
 	}
 
 	return inspection
 }
 
+// inspectIdent analyzes identifier expressions in CEL and determines if they are known resources
+// or unknown references. It handles the base identifiers in field access chains and distinguishes
+// between declared resources and unknown/internal identifiers.
```
Comment (michaelhtm) [inline internal/typesystem/celinspector/celinspector.go:247]: i think we can find a better name for this function as well...
`inspectVariable`, `inspectSymbol`..

File: internal/typesystem/celinspector/celinspector.go [hunk 4]
```diff
@@ -133,45 +193,63 @@ func (a *Inspector) inspectAst(expr *exprpb.Expr, currentPath string) Expression
 	}
 }
 
+// inspectCall analyzes function calls and method invocations within a CEL expression.
+// It tracks three types of calls:
+// 1. Custom functions (declared in Inspector initialization)
+// 2. Method calls on resources ( list.filter(...))
+// 3. Unknown functions (neither custom nor internal)
```
Comment (michaelhtm) [inline internal/typesystem/celinspector/celinspector.go:201]: `inspectCallExpression`

File: internal/typesystem/celinspector/celinspector.go (hunk 1 above)
Comment (a-hilaly) [reply internal/typesystem/celinspector/celinspector.go:137]: Great catch!

File: internal/typesystem/celinspector/celinspector.go (hunk 3 above)
Comment (a-hilaly) [reply internal/typesystem/celinspector/celinspector.go:247]: I agree there are better names to use here, however the naming choice here was to make as close to the CEL AST library (the `Ident` name comes from the CEL library 🤷 )

File: internal/typesystem/celinspector/celinspector.go (hunk 4 above)
Comment (a-hilaly) [reply internal/typesystem/celinspector/celinspector.go:201]: ditto

File: internal/typesystem/celinspector/celinspector.go (hunk 3 above)
Comment (a-hilaly) [reply internal/typesystem/celinspector/celinspector.go:247]: I think CEL libraries are going to evolve and change their naming standards with time (before stabilizing and releasing v1.0.0), we'll need to keep an eye on that and update our library accordingly.
